### PR TITLE
46 centops implement cosmosdb data layer

### LINKED ISSUE #46: CentOps: Implement CosmosDb Data Layer
The initial CentOps Admin API will have an In Memory data store.  We want to use CosmosDB as is the plan for this project.

**Depends on:** https://github.com/buerokratt/Infrastructure/issues/33

### Acceptance Criteria
- [x] Implement CosmosDb Data Layer

### LINKED ISSUE #62: CentOps: CosmosDB Tests
As A CentOps engineer
I want to test the CosmosDB functionality via unit tests
So that the CosmosDB store can be verified to be running correctly

### Acceptance Criteria
- [x] Unit Tests written for CosmosDB

### PR DESCRIPTION
**Pull request by adelechurchill**
Apologies this is a rather large PR, it encapsulates issues #46 and #62

- Implemented Cosmos DB layer
- Implemented Unit tests for Cosmos DB
- Restructured unit tests for InstitutionStore and ParticipantStore

### CODE REVIEW

**Review by shelabd — APPROVED** [review]

File: src/CentOps.Api/Controllers/InstitutionController.cs [hunk 1]
```diff
@@ -51,7 +51,7 @@ public async Task<ActionResult<InstitutionResponseModel>> Post(CreateUpdateInsit
                 var institutionDTO = _mapper.Map<InstitutionDto>(institution);
                 var createdInstitution = await _store.Create(institutionDTO).ConfigureAwait(false);
                 return Created(
-                    new Uri($"/admin/institutions/{createdInstitution.Name}", UriKind.Relative),
+                    new Uri($"/admin/institutions/{createdInstitution.Id}", UriKind.Relative),
```
Comment (guypritchard) [inline src/CentOps.Api/Controllers/InstitutionController.cs:54]: Good spot

File: src/CentOps.Api/Services/CosmosDbService.cs [hunk 2]
```diff
@@ -0,0 +1,371 @@
...
+
+        async Task<bool> IModelStore<InstitutionDto>.DeleteById(string id)
+        {
+            if (string.IsNullOrEmpty(id))
+            {
+                throw new ArgumentException($"{nameof(id)} not specified.");
+            }
+
+            using (ResponseMessage responseMessage = await _container.ReadItemStreamAsync(
+                partitionKey: new PartitionKey($"institution::{id}"),
```
Comment (guypritchard) [inline src/CentOps.Api/Services/CosmosDbService.cs:63]: Don't do this now, but perhaps 'institution' can come from the InstitutionDto with a property or a constant or even nameof(InsitutionDto)?

File: src/CentOps.Api/CentOps.Api.csproj [hunk 3]
```diff
@@ -15,11 +15,13 @@
   <ItemGroup>
     <PackageReference Include="AutoMapper" Version="11.0.1" />
     <PackageReference Include="AutoMapper.Extensions.Microsoft.DependencyInjection" Version="11.0.0" />
+    <PackageReference Include="Microsoft.Azure.Cosmos" Version="3.28.0" />
     <PackageReference Include="Microsoft.CodeAnalysis.NetAnalyzers" Version="6.0.0">
       <PrivateAssets>all</PrivateAssets>
       <IncludeAssets>runtime; build; native; contentfiles; analyzers; buildtransitive</IncludeAssets>
     </PackageReference>
     <PackageReference Include="Microsoft.VisualStudio.Azure.Containers.Tools.Targets" Version="1.15.1" />
+    <PackageReference Include="Newtonsoft.Json" Version="13.0.1" />
```
Comment (guypritchard) [inline src/CentOps.Api/CentOps.Api.csproj:24]: Are we really using this?

File: src/CentOps.UnitTests/CosmosDbTests.cs [hunk 4]
```diff
@@ -0,0 +1,749 @@
...
+                .Setup(c => c.GetItemQueryIterator<InstitutionDto>(
+                    It.IsAny<QueryDefinition>(),
+                    It.IsAny<string>(),
+                    It.IsAny<QueryRequestOptions>()))
+                .Returns(feedIteratorMock.Object);
+
+            var sut = new CosmosDbService(mockClient.Object, "", "") as IInstitutionStore;
+
+            // 
+            // _ = await sut.Create(institution1).ConfigureAwait(false);
```
Comment (guypritchard) [inline src/CentOps.UnitTests/CosmosDbTests.cs]: This still needed?

File: src/CentOps.UnitTests/CosmosDbTests.cs [hunk 5]
```diff
@@ -0,0 +1,749 @@
...
+            Assert.NotNull(createdInstitution.Id);
+            Assert.Equal(institution.Name, createdInstitution.Name);
+            Assert.Equal(createdInstitution.Status, createdInstitution.Status);
+        }
+
+
+        [Fact]
+        public async Task CreateThrowsIfInstitutiontIsNull()
+        {
+            // Arrange
```
Comment (guypritchard) [inline src/CentOps.UnitTests/CosmosDbTests.cs]: Nitpick: Extra Arranging

File: src/CentOps.UnitTests/CosmosDbTests.cs [hunk 6]
```diff
@@ -0,0 +1,749 @@
...
+            var sut = new CosmosDbService(mockClient.Object, "", "") as IInstitutionStore;
+
+            var institution =
+                new InstitutionDto
+                {
+                    Name = "Test",
+                    Status = InstitutionStatusDto.Active
+                };
+
+            var mockItemResponse = new Mock<ItemResponse<InstitutionDto>>();
```
Comment (guypritchard) [inline src/CentOps.UnitTests/CosmosDbTests.cs]: I don't want to disturb the code in this commit too much - but can some of this test setup be factored out into a helper method or something?

**Review by guypritchard — APPROVED** [review]
Let's kick the tyres on this - nice work!

File: src/CentOps.Api/CentOps.Api.csproj (hunk 3 above)
Comment (adelechurchill) [reply src/CentOps.Api/CentOps.Api.csproj:24]: We currently use JsonPropertyName to link the IModels/DTO's to CosmosDB, I did do a bit of research on this and there is a dependency between Cosmos v3 and Newtonsoft Json https://github.com/Azure/azure-cosmos-dotnet-v3/issues/202#issuecomment-548147813 - It appears that using `System.Text.Json` as a CustomSerializer can still cause issues with serialisation of models. I can look further into alternatives if need be

File: src/CentOps.UnitTests/CosmosDbTests.cs (hunk 4 above)
Comment (adelechurchill) [reply src/CentOps.UnitTests/CosmosDbTests.cs]: Removed old test file - already covered by CosmosDbInstitutionStoreTests and CosmosDbParticipantStoreTests

File: src/CentOps.UnitTests/CosmosDbTests.cs (hunk 5 above)
Comment (adelechurchill) [reply src/CentOps.UnitTests/CosmosDbTests.cs]: Removed old test file - already covered by CosmosDbInstitutionStoreTests and CosmosDbParticipantStoreTests

File: src/CentOps.UnitTests/CosmosDbTests.cs (hunk 6 above)
Comment (adelechurchill) [reply src/CentOps.UnitTests/CosmosDbTests.cs]: Removed old test file - already covered by CosmosDbInstitutionStoreTests and CosmosDbParticipantStoreTests

File: src/CentOps.Api/CentOps.Api.csproj (hunk 3 above)
Comment (guypritchard) [reply src/CentOps.Api/CentOps.Api.csproj:24]: Ah okay - fair enough!  Looks like Cosmos SDK Team are catching up in V4.  Okay - ignore me.

**Pull request merged**